runelite-client/src/main/java/net/runelite/client/plugins/chatpreviousworld/PreviousWorldChatPlugin.java

```diff
@@ -0,0 +1,213 @@
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.chatpreviousworld;
+
+import com.google.common.collect.EvictingQueue;
+import com.google.common.eventbus.Subscribe;
+import com.google.inject.Inject;
+import com.google.inject.Provides;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Queue;
+import net.runelite.api.Client;
+import net.runelite.api.MessageNode;
+import net.runelite.api.events.GameStateChanged;
+import net.runelite.api.events.SetMessage;
+import net.runelite.client.chat.ChatColorType;
+import net.runelite.client.chat.ChatMessageBuilder;
+import net.runelite.client.chat.ChatMessageManager;
+import net.runelite.client.chat.QueuedMessage;
+import net.runelite.client.config.ConfigManager;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+
+@PluginDescriptor(name = "World History in chat")
+public class PreviousWorldChatPlugin extends Plugin
+{
+	private static final String WORLD_TEXT = "You've previously logged in within the following worlds:";
+	private static final String WELCOME_MESSAGE = "Welcome to RuneScape.";
+
+	@Inject
+	private Client client;
+
+	@Inject
+	private PreviousWorldChatConfig config;
+
+	@Inject
+	private ChatMessageManager chatMessageManager;
+
+	private SetMessage lastMessage;
+	private Queue<QueuedMessage> messageQueue;
+
+	@Provides
+	PreviousWorldChatConfig getConfig(ConfigManager configManager)
+	{
+		return configManager.getConfig(PreviousWorldChatConfig.class);
+	}
+
+	@Subscribe
+	public void onSetMessage(SetMessage message)
+	{
+		// Start sending old messages right after the welcome message, as that is most reliable source
+		// of information that chat history was reset
+		if (message.getValue().equals(WELCOME_MESSAGE))
+		{
+			QueuedMessage queuedMessage;
+
+			while ((queuedMessage = messageQueue.poll()) != null)
+			{
+				chatMessageManager.queue(queuedMessage);
+			}
+		}
+
+		// If the messsage contains our world message, we want to update it to reflect new(er) worlds.
+		if (message.getValue().contains(WORLD_TEXT))
+		{
+			editWorldMessage(message);
+		}
+		else
+		{
+			sendWorldMessage(message);
+		}
+	}
+
+	@Override
+	protected void startUp()
+	{
+		messageQueue = EvictingQueue.create(100);
+	}
+
+	@Override
+	protected void shutDown()
+	{
+		saveLatestWorld();
+		messageQueue.clear();
```

Comment:
Why save world and clear message queue? Shouldn't `messageQueue = null` be enough for shutdown?